File: mcp_server/im/dispatcher.py
from __future__ import annotations

import random
import uuid
from typing import Any

from mock_data.active_orders import get_im_order, save_im_order
from mock_data.go_to_items import items_for_address
from mock_data.instamart_catalog import PRODUCTS, PRODUCTS_BY_SPIN
from mock_data.pune_addresses import ADDRESSES, get_address_by_id, public_address
from mcp_server.common import get_mock_scenario, get_param, simulated_latency_jitter_ms, tool_log
from mcp_server.food.dispatcher import handle_get_addresses as food_get_addresses
from mcp_server.session_store import clear_im_cart, get_session, resolve_session_id
from mcp_server.tool_aliases import resolve_method


def _error(code: str, message: str) -> tuple[bool, None, dict[str, Any]]:
    return False, None, {"code": code, "message": message}


def handle_get_addresses(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    return food_get_addresses(params)


def handle_search_products(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    address_id = get_param(params, "addressId", "address_id")
    q = str(get_param(params, "query", "q") or "").lower().strip()
    rows = PRODUCTS
    if q:
        rows = [
            p for p in PRODUCTS
            if q in p["name"].lower() or q in p.get("category", "").lower()
        ] or PRODUCTS
    out = []
    for p in rows[:20]:
        # Get price from first variant or top-level price field
        variants = p.get("variants", [])
        price_inr = None
        if variants:
            price_inr = variants[0].get("price_inr")
        if price_inr is None:
            price_inr = p.get("price_inr")
        out.append({
            "product_id": p["product_id"],
            "name": p["name"],
            "category": p["category"],
            "price_inr": price_inr,
            "variants": variants,
        })
    data = {"products": out, "query": q or None, "addressId": address_id}
    tool_log("im", "search_products", params or {}, f"{len(out)} products")
    return True, data, None


def handle_your_go_to_items(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    address_id = str(get_param(params, "addressId", "address_id") or "addr_kp_001")
    party = bool(get_param(params, "party", "partyMode"))
    items = items_for_address(address_id, party=party)
    data = {"products": items, "addressId": address_id}
    tool_log("im", "your_go_to_items", params or {}, f"{len(items)} go-to items")
    return True, data, None


def handle_add_to_cart(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    sid = resolve_session_id(params)
    session = get_session(sid)
    addr = str(get_param(params, "selectedAddressId", "addressId", "address_id") or session.im.selected_address_id or "")
    raw_items = get_param(params, "items", "lines")
    if not isinstance(raw_items, list) or not raw_items:
        return _error("VALIDATION", "items must be a non-empty list")

    if session.im.selected_address_id and addr and session.im.selected_address_id != addr and session.im.lines:
        clear_im_cart(sid)
        session = get_session(sid)

    lines: list[dict[str, Any]] = []
    subtotal = 0
    for row in raw_items:
        if not isinstance(row, dict):
            continue
        spin = str(row.get("spinId") or row.get("spin_id") or "").strip()
        pid = str(row.get("product_id") or row.get("productId") or "").strip()
        qty = int(row.get("qty", row.get("quantity", 1)) or 1)
        sku = PRODUCTS_BY_SPIN.get(spin) if spin else None
        if not sku and pid:
            sku = next((p for p in PRODUCTS if p["product_id"] == pid), None)
            if sku and sku.get("variants"):
                spin = sku["variants"][0]["spinId"]
                sku = PRODUCTS_BY_SPIN.get(spin)
        if not sku:
            return _error("VALIDATION", f"Unknown spinId/product: {spin or pid}")
        unit = int(sku["price_inr"])
        lt = unit * qty
        subtotal += lt
        lines.append({
            "spinId": spin,
            "product_id": sku["product_id"],
            "name": sku["name"],
            "qty": qty,
            "quantity": qty,
            "unit_price_inr": unit,
            "line_total_inr": lt,
        })

    session.im.selected_address_id = addr or session.im.selected_address_id or "addr_kp_001"
    session.im.lines = lines
    cart_id = f"cart_im_{sid}"
    blob = {"cart_id": cart_id, "lines": lines, "items": lines, "subtotal_inr": subtotal, "total": subtotal}
    tool_log("im", "add_to_cart", params or {}, f"cart={cart_id} subtotal INR {subtotal}")
    return True, blob, None


def handle_get_cart(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    if get_mock_scenario() == "cart_expired":
        return _error("CART_EXPIRED", "Instamart cart expired")
    sid = resolve_session_id(params)
    session = get_session(sid)
    subtotal = sum(int(l.get("line_total_inr", 0)) for l in session.im.lines)
    if subtotal > 0 and subtotal < 99:
        return _error("MIN_ORDER_NOT_MET", "Minimum order is ₹99 for Instamart")
    data = {
        "items": session.im.lines,
        "selectedAddressId": session.im.selected_address_id,
        "subtotal_inr": subtotal,
        "total": subtotal,
        "bill": {"itemTotal": subtotal, "delivery": 0, "grandTotal": subtotal},
        "availablePaymentMethods": ["COD"],
    }
    tool_log("im", "get_cart", params or {}, f"{len(session.im.lines)} items")
    return True, data, None


def handle_clear_cart(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    sid = resolve_session_id(params)
    clear_im_cart(sid)
    tool_log("im", "clear_cart", params or {}, "cleared")
    return True, {"cleared": True}, None


def handle_checkout(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    sid = resolve_session_id(params)
    session = get_session(sid)
    cart_id = get_param(params, "cartId", "cart_id") or f"cart_im_{sid}"
    if not session.im.lines:
        return _error("VALIDATION", "Cart is empty")
    subtotal = sum(int(l.get("line_total_inr", 0)) for l in session.im.lines)
    if subtotal < 99:
        return _error("MIN_ORDER_NOT_MET", "Minimum order is ₹99")
    oid = f"IM_ORD_{uuid.uuid4().hex[:8].upper()}"
    eta = random.randint(18, 45)
    msg = "Swiggy Instamart order placed successfully. Packed and out for delivery."
    data = {
        "order_id": oid,
        "orderId": oid,
        "cart_id": cart_id,
        "eta_mins": eta,
        "subtotal_inr": subtotal,
        "message": msg,
    }
    save_im_order(data)
    clear_im_cart(sid)
    tool_log("im", "checkout", params or {}, f"placed {oid}")
    return True, data, None


def handle_get_orders(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    from mock_data.active_orders import list_im_orders
    simulated_latency_jitter_ms()
    orders = list_im_orders()
    data = {"orders": orders}
    tool_log("im", "get_orders", params or {}, f"{len(orders)} orders")
    return True, data, None


def handle_get_order_details(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    oid = str(get_param(params, "orderId", "order_id") or "")
    order = get_im_order(oid)
    if not order:
        return _error("NOT_FOUND", f"Order {oid} not found")
    return True, order, None


def handle_track_order(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    oid = str(get_param(params, "orderId", "order_id") or "")
    order = get_im_order(oid)
    if not order:
        return _error("NOT_FOUND", f"Order {oid} not found")
    data = {"orderId": oid, "status": "OUT_FOR_DELIVERY", "eta_mins": order.get("eta_mins", 20)}
    return True, data, None


def handle_create_address(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    aid = f"addr_new_{uuid.uuid4().hex[:6]}"
    data = {"addressId": aid, "message": "Address created (mock)"}
    tool_log("im", "create_address", params or {}, aid)
    return True, data, None


def handle_delete_address(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    aid = str(get_param(params, "addressId", "address_id") or "")
    if not get_address_by_id(aid):
        return _error("NOT_FOUND", f"Address {aid} not found")
    tool_log("im", "delete_address", params or {}, aid)
    return True, {"deleted": aid}, None


def handle_report_error(params: dict[str, Any]) -> tuple[bool, dict | None, dict | None]:
    simulated_latency_jitter_ms()
    return True, {"reportLink": "mailto:builders@swiggy.in", "summary": "mock report"}, None


_TOOLS = {
    "get_addresses": handle_get_addresses,
    "search_products": handle_search_products,
    "your_go_to_items": handle_your_go_to_items,
    "add_to_cart": handle_add_to_cart,
    "get_cart": handle_get_cart,
    "clear_cart": handle_clear_cart,
    "checkout": handle_checkout,
    "get_orders": handle_get_orders,
    "get_order_details": handle_get_order_details,
    "track_order": handle_track_order,
    "create_address": handle_create_address,
    "delete_address": handle_delete_address,
    "report_error": handle_report_error,
}


def invoke(method: str | None, params: dict[str, Any] | None) -> tuple[bool, Any, dict[str, Any] | None]:
    if not method or not isinstance(method, str):
        return _error("VALIDATION", "method is required")
    resolved = resolve_method("im", method)
    fn = _TOOLS.get(resolved)
    if not fn:
        return _error("UNKNOWN_METHOD", f"Unknown tool: {method}")
    return fn(dict(params or {}))

File: frontend/lib/mcp-mock/dispatcher.ts
import {
  ADDRESSES,
  DINEOUT_VENUES,
  FOOD_COUPONS,
  GO_TO_ITEMS,
  MENUS,
  PRODUCTS,
  PRODUCTS_BY_SPIN,
  RESTAURANTS,
  SLOT_LABELS,
  getAddressById,
  publicAddress,
} from "./data";
import { getSession, resolveSessionId } from "./session-store";

export type McpVertical = "food" | "im" | "dineout";

const ALIASES: Record<string, string> = {
  get_restaurant_menu: "get_menu",
  update_food_cart: "add_to_cart",
  place_food_order: "place_order",
  update_cart: "add_to_cart",
  search_restaurants_dineout: "search_restaurants",
  get_available_slots: "check_availability",
};

function resolveMethod(method: string): string {
  return ALIASES[method] ?? method;
}

function err(code: string, message: string) {
  return { success: false as const, error: { code, message } };
}

function ok(data: unknown) {
  return { success: true as const, data };
}

function pickEta(r: (typeof RESTAURANTS)[0]) {
  return Math.floor((r.eta_mins_min + r.eta_mins_max) / 2);
}

function foodCartView(sid: string, addressId?: string) {
  const s = getSession(sid);
  const subtotal = s.food.lines.reduce((n, l) => n + l.line_total_inr, 0);
  const discount = s.food.coupon_discount ?? 0;
  const delivery = subtotal > 0 ? 40 : 0;
  const total = Math.max(0, subtotal + delivery - discount);
  const rest = RESTAURANTS.find((r) => r.restaurant_id === s.food.restaurant_id);
  return {
    restaurantId: s.food.restaurant_id || undefined,
    restaurantName: s.food.restaurant_name || rest?.name,
    addressId: addressId || s.food.address_id || undefined,
    items: s.food.lines,
    subtotal_inr: subtotal,
    deliveryCharge: delivery,
    total,
    availablePaymentMethods: ["COD"],
    offers: { coupon_applied: s.food.coupon_code ?? null },
  };
}

function imCartView(sid: string) {
  const s = getSession(sid);
  const subtotal = s.im.lines.reduce((n, l) => n + l.line_total_inr, 0);
  const delivery = subtotal > 0 ? 25 : 0;
  const grand = subtotal + delivery;
  return {
    cart_id: `cart_im_${sid}`,
    items: s.im.lines,
    selectedAddressId: s.im.address_id || undefined,
    subtotal_inr: subtotal,
    total: grand,
    bill: { itemTotal: subtotal, delivery, grandTotal: grand },
    availablePaymentMethods: ["COD"],
  };
}

function invokeFood(method: string, params: Record<string, unknown>) {
  const m = resolveMethod(method);
  const sid = resolveSessionId(params);

  if (m === "get_addresses") {
    return ok({ addresses: ADDRESSES.map(publicAddress) });
  }
  if (m === "search_restaurants") {
    const addressId = String(params.addressId ?? params.address_id ?? "");
    if (!addressId) return err("VALIDATION", "addressId is required");
    const addr = getAddressById(addressId);
    if (!addr) return err("NOT_FOUND", `Unknown addressId: ${addressId}`);
    const q = String(params.query ?? params.q ?? "").toLowerCase();
    let rows = [...RESTAURANTS];
    if (q) {
      rows = rows.filter(
        (r) =>
          r.name.toLowerCase().includes(q) ||
          r.cuisines.some((c) => c.toLowerCase().includes(q))
      );
    }
    return ok({
      addressId,
      area: addr.area,
      restaurants: rows.map((r) => ({
        restaurant_id: r.restaurant_id,
        name: r.name,
        rating: r.rating,
        eta_mins: pickEta(r),
        distanceKm: r.distance_km,
        availabilityStatus: r.availability_status,
        cuisines: r.cuisines,
        tag: r.tag,
      })),
      nextOffset: rows.length,
    });
  }
  if (m === "get_menu") {
    const rid = String(params.restaurantId ?? params.restaurant_id ?? "");
    const menu = MENUS[rid];
    if (!menu) return err("NOT_FOUND", `No menu for ${rid}`);
    return ok(menu);
  }
  if (m === "add_to_cart") {
    const rid = String(params.restaurantId ?? params.restaurant_id ?? "");
    const addressId = String(params.addressId ?? params.address_id ?? "");
    const raw = (params.lines ?? params.items ?? params.cartItems) as unknown[];
    if (!rid || !Array.isArray(raw) || !raw.length) {
      return err("VALIDATION", "restaurantId and lines required");
    }
    const s = getSession(sid);
    if (s.food.restaurant_id && s.food.restaurant_id !== rid && s.food.lines.length) {
      s.food = { restaurant_id: "", restaurant_name: "", address_id: "", lines: [] };
    }
    const menu = MENUS[rid];
    const prices: Record<string, { name: string; price: number }> = {};
    for (const cat of menu?.categories ?? []) {
      for (const it of cat.items) {
        prices[it.item_id] = { name: it.name, price: it.price_inr };
      }
    }
    const rest = RESTAURANTS.find((r) => r.restaurant_id === rid);
    s.food.restaurant_id = rid;
    s.food.restaurant_name = rest?.name ?? rid;
    s.food.address_id = addressId;
    s.food.lines = raw.map((row) => {
      const r = row as Record<string, unknown>;
      const itemId = String(r.item_id ?? r.itemId ?? "");
      const qty = Number(r.qty ?? r.quantity ?? 1);
      const p = prices[itemId]?.price ?? 199;
      return {
        item_id: itemId,
        name: prices[itemId]?.name ?? itemId,
        qty,
        unit_price_inr: p,
        line_total_inr: p * qty,
      };
    });
    return ok({ ...foodCartView(sid, addressId), message: "Cart updated" });
  }
  if (m === "get_food_cart") {
    return ok(foodCartView(sid, String(params.addressId ?? params.address_id ?? "")));
  }
  if (m === "flush_food_cart") {
    getSession(sid).food = { restaurant_id: "", restaurant_name: "", address_id: "", lines: [] };
    return ok({ cleared: true });
  }
  if (m === "fetch_food_coupons") {
    return ok({ coupons: FOOD_COUPONS });
  }
  if (m === "apply_food_coupon") {
    const code = String(params.code ?? params.couponCode ?? "");
    const coupon = FOOD_COUPONS.find((c) => c.code === code);
    if (!coupon) return err("NOT_FOUND", `Unknown coupon ${code}`);
    const s = getSession(sid);
    s.food.coupon_code = code;
    s.food.coupon_discount = coupon.discount_inr;
    return ok({ code, discount_inr: coupon.discount_inr });
  }
  if (m === "place_order") {
    const view = foodCartView(sid);
    if (!view.items?.length) return err("VALIDATION", "Cart is empty");
    if ((view.total ?? 0) >= 1000) return err("LIMIT", "Mock cap: orders above ₹999 need app handoff");
    const orderId = `fd-ord-${Date.now()}`;
    const s = getSession(sid);
    const order = { orderId, status: "ACTIVE", ...view, eta_mins: 32, paymentMethod: "COD" };
    s.food_orders.push(order);
    s.last_food_order_id = orderId;
    s.food = { restaurant_id: "", restaurant_name: "", address_id: "", lines: [] };
    return ok({ ...order, order_id: orderId, orderId, message: "Order placed (mock)" });
  }
  if (m === "get_food_orders") {
    return ok({ orders: getSession(sid).food_orders });
  }
  if (m === "get_food_order_details") {
    const oid = String(params.orderId ?? params.order_id ?? "");
    const o = getSession(sid).food_orders.find((x) => x.orderId === oid);
    if (!o) return err("NOT_FOUND", "Order not found");
    return ok(o);
  }
  if (m === "track_food_order") {
    const oid = String(params.orderId ?? params.order_id ?? getSession(sid).last_food_order_id ?? "");
    return ok({ orderId: oid, status: "OUT_FOR_DELIVERY", eta_mins: 18, deliveryTimeSpoken: "about 18 minutes" });
  }
  if (m === "report_error") {
    return ok({ reportLink: "https://example.com/report", summary: "Mock error report filed" });
  }
  return err("UNKNOWN", `Unknown food method: ${method}`);
}

function invokeIm(method: string, params: Record<string, unknown>) {
  const m = resolveMethod(method);
  const sid = resolveSessionId(params);

  if (m === "get_addresses") return invokeFood("get_addresses", params);
  if (m === "search_products") {
    const q = String(params.query ?? params.q ?? "").toLowerCase();
    let rows = PRODUCTS;
    if (q) rows = rows.filter((p) => p.name.toLowerCase().includes(q) || p.category.includes(q));
    return ok({
      products: rows.map((p) => ({ product_id: p.product_id, name: p.name, category: p.category, variants: p.variants })),
      query: q || null,
      addressId: params.addressId,
    });
  }
  if (m === "your_go_to_items") {
    const party = Boolean(params.party ?? params.partyMode);
    const items = party
      ? [...GO_TO_ITEMS, { product_id: "im_plate_01", name: "Party plates", spinId: "spin_plate_50", price_inr: 189, category: "party" }]
      : GO_TO_ITEMS;
    return ok({
      products: items.map((g) => ({
        product_id: g.product_id,
        name: g.name,
        variants: [{ spinId: g.spinId, label: "default", price_inr: g.price_inr, inStock: true }],
      })),
      addressId: params.addressId ?? "addr_kp_001",
    });
  }
  if (m === "add_to_cart") {
    const addr = String(params.selectedAddressId ?? params.addressId ?? "");
    const raw = (params.items ?? params.lines) as unknown[];
    if (!Array.isArray(raw) || !raw.length) return err("VALIDATION", "items required");
    const s = getSession(sid);
    if (s.im.address_id && addr && s.im.address_id !== addr && s.im.lines.length) {
      s.im = { address_id: "", lines: [] };
    }
    s.im.address_id = addr;
    s.im.lines = raw.map((row) => {
      const r = row as Record<string, unknown>;
      const spinId = String(r.spinId ?? r.spin_id ?? "");
      const qty = Number(r.qty ?? r.quantity ?? 1);
      const sku = PRODUCTS_BY_SPIN[spinId];
      const price = sku?.variants[0]?.price_inr ?? 99;
      return {
        spinId,
        product_id: sku?.product_id ?? spinId,
        name: sku?.name ?? spinId,
        qty,
        unit_price_inr: price,
        line_total_inr: price * qty,
      };
    });
    return ok(imCartView(sid));
  }
  if (m === "get_cart") return ok(imCartView(sid));
  if (m === "clear_cart") {
    getSession(sid).im = { address_id: "", lines: [] };
    return ok({ cleared: true });
  }
  if (m === "checkout") {
    const view = imCartView(sid);
    if (!view.items?.length) return err("VALIDATION", "Cart is empty");
    if ((view.subtotal_inr ?? 0) < 99) return err("MIN_ORDER", "Minimum order ₹99");
    const orderId = `im-ord-${Date.now()}`;
    const s = getSession(sid);
    const order = { orderId, status: "ACTIVE", ...view, eta_mins: 25 };
    s.im_orders.push(order);
    s.last_im_order_id = orderId;
    s.im = { address_id: "", lines: [] };
    return ok({ ...order, order_id: orderId, orderId, message: "Checkout complete (mock)" });
  }
  if (m === "get_orders") return ok({ orders: getSession(sid).im_orders });
  if (m === "get_order_details") {
    const oid = String(params.orderId ?? params.order_id ?? "");
    const o = getSession(sid).im_orders.find((x) => x.orderId === oid);
    if (!o) return err("NOT_FOUND", "Order not found");
    return ok(o);
  }
  if (m === "track_order") {
    const oid = String(params.orderId ?? params.order_id ?? getSession(sid).last_im_order_id ?? "");
    return ok({ orderId: oid, status: "OUT_FOR_DELIVERY", eta_mins: 22 });
  }
  if (m === "create_address") {
    return ok({ addressId: `addr_new_${Date.now()}`, message: "Address created (mock)" });
  }
  if (m === "delete_address") {
    return ok({ deleted: params.addressId });
  }
  if (m === "report_error") {
    return ok({ reportLink: "https://example.com/report", summary: "Mock error report filed" });
  }
  return err("UNKNOWN", `Unknown im method: ${method}`);
}

function invokeDineout(method: string, params: Record<string, unknown>) {
  const m = resolveMethod(method);

  if (m === "get_saved_locations") {
    return ok({
      locations: ADDRESSES.map((a) => ({
        addressId: a.addressId,
        label: a.label,
        lat: a._latitude,
        lng: a._longitude,
        latitude: a._latitude,
        longitude: a._longitude,
      })),
    });
  }
  if (m === "search_restaurants") {
    const q = String(params.query ?? params.q ?? "").toLowerCase();
    let rows = DINEOUT_VENUES.filter((v) => v.availability === "AVAILABLE");
    if (q) {
      rows = rows.filter(
        (v) =>
          v.name.toLowerCase().includes(q) ||
          v.cuisines.some((c) => c.toLowerCase().includes(q))
      );
    }
    return ok({
      restaurants: rows.map((v) => ({
        restaurant_id: v.restaurant_id,
        name: v.name,
        rating: v.rating,
        cuisines: v.cuisines,
        area: v.area,
        costForTwo: v.costForTwo,
        availability: v.availability,
        booking_type: "TABLE",
        latitude: v._lat,
        longitude: v._lng,
      })),
    });
  }
  if (m === "check_availability") {
    const rid = String(params.restaurantId ?? params.restaurant_id ?? "");
    const guests = Number(params.guestCount ?? params.partySize ?? params.party_size ?? 2);
    const venue = DINEOUT_VENUES.find((v) => v.restaurant_id === rid);
    if (!venue) return err("NOT_FOUND", "Venue not found");
    if (venue.availability === "UNAVAILABLE") return err("UNAVAILABLE", "Walk-in only");
    const slots = SLOT_LABELS.map((label, i) => ({
      slotId: 4200 + i,
      label,
      reservationTime: Math.floor(Date.now() / 1000) + (i + 1) * 3600,
      guestCount: guests,
      band: "dinner",
      deals: [{ slotId: 4200 + i, itemId: `${rid}-ticket_${4200 + i}`, isFree: true, bookingPrice: 0, title: "Free table reservation" }],
    }));
    return ok({ restaurant_id: rid, name: venue.name, party_size: guests, guestCount: guests, available: true, slots });
  }
  if (m === "book_table") {
    const rid = String(params.restaurantId ?? params.restaurant_id ?? "");
    const venue = DINEOUT_VENUES.find((v) => v.restaurant_id === rid);
    const bookingId = `bk-${Date.now()}`;
    const booking = {
      booking_id: bookingId,
      bookingId,
      restaurant_id: rid,
      venue_name: venue?.name ?? "Restaurant",
      guests: Number(params.partySize ?? params.guestCount ?? 2),
      slot: params.slot ?? "20:00",
      slotId: params.slotId,
      itemId: params.itemId,
      status: "CONFIRMED",
      confirmation_message: `Table confirmed at ${venue?.name ?? "venue"} (mock).`,
    };
    const sid = resolveSessionId(params);
    const s = getSession(sid);
    s.bookings.push(booking);
    s.last_booking_id = bookingId;
    return ok(booking);
  }
  if (m === "get_booking_status") {
    const bid = String(params.bookingId ?? params.booking_id ?? "");
    const sid = resolveSessionId(params);
    const b = getSession(sid).bookings.find((x) => x.bookingId === bid);
    if (!b) return err("NOT_FOUND", "Booking not found");
    return ok(b);
  }
  if (m === "create_cart") {
    return ok({ cartId: `do-cart-${Date.now()}`, message: "Dineout cart stub" });
  }
  if (m === "report_error") {
    return ok({ reportLink: "https://example.com/report", summary: "Mock error report filed" });
  }
  return err("UNKNOWN", `Unknown dineout method: ${method}`);
}

export function invokeMockMcp(vertical: McpVertical, method: string, params: Record<string, unknown> = {}) {
  if (vertical === "food") return invokeFood(method, params);
  if (vertical === "im") return invokeIm(method, params);
  return invokeDineout(method, params);
}

export const ALL_MCP_METHODS: Record<McpVertical, string[]> = {
  food: ["get_addresses", "search_restaurants", "get_menu", "add_to_cart", "get_food_cart", "flush_food_cart", "fetch_food_coupons", "apply_food_coupon", "place_order", "get_food_orders", "get_food_order_details", "track_food_order", "report_error"],
  im: ["get_addresses", "search_products", "your_go_to_items", "add_to_cart", "get_cart", "clear_cart", "checkout", "get_orders", "get_order_details", "track_order", "create_address", "delete_address", "report_error"],
  dineout: ["get_saved_locations", "search_restaurants", "check_availability", "book_table", "get_booking_status", "create_cart", "report_error"],
};
